File: wisp/multi_agent/dag.py
"""Task DAG — directed acyclic graph of subagent tasks with topological scheduling.

Level-by-level parallel execution. Each level's tasks run concurrently
up to max_parallelism. Dependencies must complete before dependents start.
"""

from __future__ import annotations

import asyncio
import logging
import time
from collections import defaultdict, deque
from dataclasses import dataclass, field
from typing import Any, Callable

logger = logging.getLogger(__name__)


@dataclass
class TaskNode:
    """A node in the task DAG.

    Each node wraps a SubagentContract or callable and declares
    dependencies that must complete before it can execute.
    """

    name: str
    task: Any  # SubagentContract or callable
    dependencies: list[str] = field(default_factory=list)
    metadata: dict = field(default_factory=dict)

    def __hash__(self) -> int:
        return hash(self.name)


@dataclass
class TaskDAG:
    """A directed acyclic graph of tasks.

    Nodes are added with named dependencies. The DAG validates
    acyclicity before execution.
    """

    nodes: dict[str, TaskNode] = field(default_factory=dict)
    _edges: dict[str, set[str]] = field(default_factory=lambda: defaultdict(set))
    _reverse_edges: dict[str, set[str]] = field(default_factory=lambda: defaultdict(set))

    def add_node(self, node: TaskNode) -> None:
        if node.name in self.nodes:
            raise ValueError(f"Duplicate node name: {node.name}")
        self.nodes[node.name] = node
        if node.name not in self._edges:
            self._edges[node.name] = set()
        if node.name not in self._reverse_edges:
            self._reverse_edges[node.name] = set()
        # Process declared dependencies
        for dep in node.dependencies:
            self.add_edge(dep, node.name)

    def add_edge(self, from_node: str, to_node: str) -> None:
        """Declare that 'to_node' depends on 'from_node'.

        Nodes must exist by validation time, not necessarily at edge creation.
        """
        self._edges[from_node].add(to_node)
        self._reverse_edges[to_node].add(from_node)

    def dependencies_of(self, name: str) -> set[str]:
        """Nodes that must complete before 'name' can run."""
        return set(self._reverse_edges.get(name, set()))

    def dependents_of(self, name: str) -> set[str]:
        """Nodes that depend on 'name'."""
        return set(self._edges.get(name, set()))

    def roots(self) -> list[str]:
        """Nodes with no dependencies."""
        return [name for name in self.nodes if not self.dependencies_of(name)]

    def validate(self) -> list[str]:
        """Check for cycles. Returns list of errors (empty = valid)."""
        errors: list[str] = []

        # Check all referenced dependencies exist
        for node in self.nodes.values():
            for dep in node.dependencies:
                if dep not in self.nodes:
                    errors.append(f"Node '{node.name}' depends on unknown '{dep}'")

        # Cycle detection via Kahn's algorithm
        in_degree = {name: len(self.dependencies_of(name)) for name in self.nodes}
        q = deque([name for name, deg in in_degree.items() if deg == 0])
        visited = 0

        while q:
            current = q.popleft()
            visited += 1
            for dependent in self.dependents_of(current):
                in_degree[dependent] -= 1
                if in_degree[dependent] == 0:
                    q.append(dependent)

        if visited != len(self.nodes):
            remaining = [name for name, deg in in_degree.items() if deg > 0]
            errors.append(f"Cycle detected involving: {', '.join(remaining)}")

        return errors

    def topological_levels(self) -> list[list[str]]:
        """Group nodes into levels for parallel execution.

        Level 0 = roots (no deps). Level N = nodes whose deps are all in levels < N.
        """
        in_degree = {name: len(self.dependencies_of(name)) for name in self.nodes}
        remaining = dict(in_degree)
        levels: list[list[str]] = []

        while remaining:
            current_level = sorted(
                name for name, deg in remaining.items() if deg == 0
            )
            if not current_level:
                break  # Cycle — caller should validate first
            levels.append(current_level)
            for name in current_level:
                del remaining[name]
                for dependent in self.dependents_of(name):
                    if dependent in remaining:
                        remaining[dependent] -= 1

        return levels


@dataclass
class DAGResult:
    """Result of executing a task DAG."""

    node_results: dict[str, Any]  # name → SubagentResult
    level_order: list[list[str]]
    total_elapsed: float
    success: bool = True
    errors: list[str] = field(default_factory=list)


@dataclass
class DAGScheduler:
    """Execute a TaskDAG level-by-level with parallel execution within each level.

    Each level's tasks run concurrently up to max_parallelism.
    """

    max_parallelism: int = 4
    timeout_per_node: float = 300.0

    async def execute(
        self,
        dag: TaskDAG,
        executor: Callable[[Any], Any],
    ) -> DAGResult:
        """Execute all nodes in topological order.

        Args:
            dag: The task DAG to execute.
            executor: Async callable that takes a TaskNode and returns a result.
                      Called as: await executor(node)

        Returns:
            DAGResult with per-node results and timing.
        """
        errors = dag.validate()
        if errors:
            return DAGResult(
                node_results={},
                level_order=[],
                total_elapsed=0.0,
                success=False,
                errors=errors,
            )

        levels = dag.topological_levels()
        all_results: dict[str, Any] = {}
        start = time.monotonic()

        semaphore = asyncio.Semaphore(self.max_parallelism)

        async def _run_node(node: TaskNode) -> None:
            async with semaphore:
                try:
                    result = await asyncio.wait_for(
                        executor(node),
                        timeout=self.timeout_per_node,
                    )
                    all_results[node.name] = result
                except asyncio.TimeoutError:
                    logger.error("Node %s timed out after %.1fs", node.name, self.timeout_per_node)
                    all_results[node.name] = _timeout_result(node.name)
                except Exception as exc:
                    logger.exception("Node %s failed", node.name)
                    all_results[node.name] = _error_result(node.name, str(exc))

        for level_idx, level in enumerate(levels):
            logger.info("DAG level %d: %s", level_idx, level)
            tasks = [asyncio.create_task(_run_node(dag.nodes[name])) for name in level]
            await asyncio.gather(*tasks)

        elapsed = time.monotonic() - start
        success = all(
            getattr(r, "success", True) for r in all_results.values()
        )

        return DAGResult(
            node_results=all_results,
            level_order=levels,
            total_elapsed=elapsed,
            success=success,
        )


def _timeout_result(name: str) -> Any:
    """Create a timeout error result compatible with SubagentResult."""
    return _make_fallback(name, False, "Timed out", "timeout")


def _error_result(name: str, error: str) -> Any:
    """Create an error result compatible with SubagentResult."""
    return _make_fallback(name, False, f"Error: {error}", error)


def _make_fallback(name: str, success: bool, output: str, error: str) -> Any:
    """Create a result-like object without importing SubagentResult.

    Uses duck-typing so callers can check .success, .output, .error.
    """
    return type("_FallbackResult", (), {
        "task_id": name,
        "success": success,
        "output": output,
        "error": error,
        "elapsed_seconds": 0.0,
    })()

File: wisp-ts/src/multi_agent/dag.ts
/** Task DAG — directed acyclic graph of subagent tasks with topological scheduling. */

export class TaskNode {
  name: string;
  task: unknown;
  dependencies: string[];
  metadata: Record<string, unknown>;

  constructor(name: string, task: unknown, dependencies: string[] = [], metadata: Record<string, unknown> = {}) {
    this.name = name;
    this.task = task;
    this.dependencies = dependencies;
    this.metadata = metadata;
  }
}

export class TaskDAG {
  nodes = new Map<string, TaskNode>();
  private _edges = new Map<string, Set<string>>();
  private _reverseEdges = new Map<string, Set<string>>();

  addNode(node: TaskNode): void {
    if (this.nodes.has(node.name)) throw new Error(`Duplicate node name: ${node.name}`);
    this.nodes.set(node.name, node);
    if (!this._edges.has(node.name)) this._edges.set(node.name, new Set());
    if (!this._reverseEdges.has(node.name)) this._reverseEdges.set(node.name, new Set());
    for (const dep of node.dependencies) {
      this.addEdge(dep, node.name);
    }
  }

  addEdge(fromNode: string, toNode: string): void {
    if (!this._edges.has(fromNode)) this._edges.set(fromNode, new Set());
    if (!this._reverseEdges.has(toNode)) this._reverseEdges.set(toNode, new Set());
    this._edges.get(fromNode)!.add(toNode);
    this._reverseEdges.get(toNode)!.add(fromNode);
  }

  dependenciesOf(name: string): Set<string> {
    return new Set(this._reverseEdges.get(name) ?? []);
  }

  dependentsOf(name: string): Set<string> {
    return new Set(this._edges.get(name) ?? []);
  }

  roots(): string[] {
    return Array.from(this.nodes.keys()).filter((name) => this.dependenciesOf(name).size === 0);
  }

  validate(): string[] {
    const errors: string[] = [];
    for (const node of this.nodes.values()) {
      for (const dep of node.dependencies) {
        if (!this.nodes.has(dep)) errors.push(`Node '${node.name}' depends on unknown '${dep}'`);
      }
    }

    // Kahn's algorithm
    const inDegree = new Map<string, number>();
    for (const name of this.nodes.keys()) {
      inDegree.set(name, this.dependenciesOf(name).size);
    }

    const q = Array.from(inDegree.entries()).filter(([, deg]) => deg === 0).map(([name]) => name);
    let visited = 0;
    let head = 0;

    while (head < q.length) {
      const current = q[head++];
      visited++;
      for (const dependent of this.dependentsOf(current)) {
        const newDeg = (inDegree.get(dependent) ?? 0) - 1;
        inDegree.set(dependent, newDeg);
        if (newDeg === 0) q.push(dependent);
      }
    }

    if (visited !== this.nodes.size) {
      const remaining = Array.from(inDegree.entries()).filter(([, deg]) => deg > 0).map(([name]) => name);
      errors.push(`Cycle detected involving: ${remaining.join(", ")}`);
    }

    return errors;
  }

  topologicalLevels(): string[][] {
    const remaining = new Map<string, number>();
    for (const name of this.nodes.keys()) {
      remaining.set(name, this.dependenciesOf(name).size);
    }

    const levels: string[][] = [];
    while (remaining.size > 0) {
      const currentLevel = Array.from(remaining.entries())
        .filter(([, deg]) => deg === 0)
        .map(([name]) => name)
        .sort();
      if (currentLevel.length === 0) break;
      levels.push(currentLevel);
      for (const name of currentLevel) {
        remaining.delete(name);
        for (const dependent of this.dependentsOf(name)) {
          if (remaining.has(dependent)) {
            remaining.set(dependent, (remaining.get(dependent) ?? 0) - 1);
          }
        }
      }
    }
    return levels;
  }
}

export class DAGResult {
  nodeResults = new Map<string, unknown>();
  levelOrder: string[][] = [];
  totalElapsed = 0;
  success = true;
  errors: string[] = [];

  constructor(init?: Partial<DAGResult>) {
    if (init) Object.assign(this, init);
  }
}

export class DAGScheduler {
  maxParallelism: number;
  timeoutPerNode: number;

  constructor(maxParallelism = 4, timeoutPerNode = 300) {
    this.maxParallelism = maxParallelism;
    this.timeoutPerNode = timeoutPerNode;
  }

  async execute(dag: TaskDAG, executor: (node: TaskNode) => Promise<unknown>): Promise<DAGResult> {
    const errors = dag.validate();
    if (errors.length > 0) {
      return new DAGResult({ success: false, errors });
    }

    const levels = dag.topologicalLevels();
    const allResults = new Map<string, unknown>();
    const start = performance.now();

    const semaphore = this._makeSemaphore(this.maxParallelism);

    const _runNode = async (node: TaskNode): Promise<void> => {
      const release = await semaphore.acquire();
      try {
        const result = await this._withTimeout(executor(node), this.timeoutPerNode * 1000);
        allResults.set(node.name, result);
      } catch (exc) {
        const errorMsg = exc instanceof Error ? exc.message : String(exc);
        allResults.set(node.name, { success: false, error: errorMsg, output: `Error: ${errorMsg}` });
      } finally {
        release();
      }
    };

    for (const level of levels) {
      const tasks = level.map((name) => _runNode(dag.nodes.get(name)!));
      await Promise.all(tasks);
    }

    const elapsed = (performance.now() - start) / 1000;
    const success = Array.from(allResults.values()).every((r) => {
      if (r && typeof r === "object") return (r as Record<string, unknown>).success !== false;
      return true;
    });

    return new DAGResult({ nodeResults: allResults, levelOrder: levels, totalElapsed: elapsed, success });
  }

  private _makeSemaphore(n: number): { acquire(): Promise<() => void> } {
    let count = n;
    const queue: Array<{ resolve: (release: () => void) => void }> = [];
    return {
      acquire(): Promise<() => void> {
        if (count > 0) {
          count--;
          return Promise.resolve(() => {
            count++;
            const next = queue.shift();
            if (next) { count--; next.resolve(() => { count++; }); }
          });
        }
        return new Promise((resolve) => queue.push({ resolve }));
      },
    };
  }

  private async _withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error("timeout")), ms);
      promise.then(
        (v) => { clearTimeout(timer); resolve(v); },
        (e) => { clearTimeout(timer); reject(e); }
      );
    });
  }
}
